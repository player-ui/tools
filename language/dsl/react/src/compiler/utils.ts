import React from "react";
import type { DefaultCompilerContentType } from "./types";

/** Basic way of identifying the type of file based on the default content export */
export const fingerprintContent = (
  content: unknown,
  filename?: string,
): DefaultCompilerContentType | undefined => {
  if (content !== null || content !== undefined) {
    if (React.isValidElement(content as any)) {
      return "view";
    }

    if (typeof content === "object" && "navigation" in (content as any)) {
      return "flow";
    }

    if (!filename || filename.includes("schema")) {
      return "schema";
    }
  }
};
