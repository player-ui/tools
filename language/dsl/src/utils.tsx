import React from 'react';
import {
  isTemplateStringInstance,
  TemplateStringComponent,
} from './string-templates';

/** Get an array version of the value */
export function toArray<T>(val: T | Array<T>): Array<T> {
  return Array.isArray(val) ? val : [val];
}

/** Create a component version  */
export function toJsonElement(value: any, index?: number): React.ReactElement {
  const keyProp = index === undefined ? null : { key: index };

  if (Array.isArray(value)) {
    return (
      <array {...keyProp}>
        {value.map((item, idx) => toJsonElement(item, idx))}
      </array>
    );
  }

  /** Allow users to pass in BindingTemplateInstance and ExpressionTemplateInstance directly without turning them into strings first */
  if (isTemplateStringInstance(value)) {
    return <value {...keyProp}>{value.toRefString()}</value>;
  }

  if (typeof value === 'object' && value !== null) {
    return (
      <obj {...keyProp}>
        {Object.keys(value).map((key) => (
          <property key={key} name={key}>
            {toJsonElement(value[key])}
          </property>
        ))}
      </obj>
    );
  }

  return <value {...keyProp} value={value} />;
}

/** Create a fragment for the properties */
export function toJsonProperties(value: Record<string, any>) {
  return Object.keys(value).map((key) => (
    <property key={key} name={key}>
      {toJsonElement(value[key])}
    </property>
  ));
}

/** Create a text asset if needed */
export function normalizeText(options: {
  /** The current node */
  node: React.ReactNode;

  /** A component to render a text asset */
  TextComp?: React.ComponentType<React.PropsWithChildren>;
}): React.ReactNode {
  const { node, TextComp } = options;

  const nodeArr = React.Children.toArray(node);

  if (
    nodeArr.every(
      (n) => React.isValidElement(n) && n.type !== TemplateStringComponent
    )
  ) {
    return node;
  }

  if (TextComp) {
    return <TextComp>{nodeArr}</TextComp>;
  }

  throw new Error(
    `Tried to convert node to Text Asset, but no Component was supplied.`
  );
}

/** Create a collection if needed */
export function normalizeToCollection(options: {
  /** the node to look at */
  node: React.ReactNode;

  /** A Text asset */
  TextComp?: React.ComponentType;

  /** A collection asset */
  CollectionComp?: React.ComponentType<React.PropsWithChildren>;
}) {
  const { node, CollectionComp } = options;

  if (
    React.Children.count(node) > 1 &&
    React.Children.toArray(node).every((n) => typeof n !== 'string')
  ) {
    if (!CollectionComp) {
      throw new Error(
        `Tried to convert array to a collection asset, but no Component was given.`
      );
    }

    return <CollectionComp>{node}</CollectionComp>;
  }

  return normalizeText({ ...options, node });
}
