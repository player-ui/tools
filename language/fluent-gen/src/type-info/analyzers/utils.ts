/** Structured analysis error reporting. */
export interface AnalysisError {
  analyzer: string;
  property: string;
  typeText: string;
  reason: string;
  context?: Record<string, unknown>;
}

/** Logs analysis warnings with consistent formatting and optional context. */
export function logAnalysisWarning(
  analyzer: string,
  message: string,
  context?: Record<string, unknown>,
): void {
  if (process.env.NODE_ENV === "development") {
    console.warn(`[${analyzer}] ${message}`, context ? context : "");
  }
}

/** Logs analysis errors with structured information. */
export function logAnalysisError(error: AnalysisError): void {
  if (process.env.NODE_ENV === "development") {
    console.error(
      `[${error.analyzer}] Failed to analyze property "${error.property}" (${error.typeText}): ${error.reason}`,
      error.context,
    );
  }
}

/** Safely executes an analysis function with error handling. */
export function safeAnalyze<T>({
  analyzer,
  property,
  typeText,
  propertyFn,
  fallback,
}: {
  analyzer: string;
  property: string;
  typeText: string;
  propertyFn: () => T;
  fallback: T;
}): T {
  try {
    return propertyFn();
  } catch (error) {
    logAnalysisError({
      analyzer,
      property,
      typeText,
      reason: error instanceof Error ? error.message : "Unknown error",
      context: { error },
    });
    return fallback;
  }
}
