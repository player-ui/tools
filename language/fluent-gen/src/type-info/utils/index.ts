import {
  Node,
  TypeNode,
  TypeReferenceNode,
  InterfaceDeclaration,
  TypeAliasDeclaration,
} from "ts-morph";
import type { ExtractorContext } from "../core/ExtractorContext.js";
import { ARRAY_TYPE_NAMES, UTILITY_TYPES } from "./TypeGuards.js";

/**
 * Extract string literal values from a union type or single literal.
 */
export function extractStringLiteralUnion(typeNode: TypeNode): string[] {
  const literals: string[] = [];

  if (Node.isUnionTypeNode(typeNode)) {
    for (const unionType of typeNode.getTypeNodes()) {
      if (Node.isLiteralTypeNode(unionType)) {
        const literal = unionType.getLiteral();
        if (Node.isStringLiteral(literal)) {
          literals.push(literal.getLiteralValue());
        }
      }
    }
  } else if (Node.isLiteralTypeNode(typeNode)) {
    const literal = typeNode.getLiteral();
    if (Node.isStringLiteral(literal)) {
      literals.push(literal.getLiteralValue());
    }
  }

  return literals;
}

/**
 * Find an interface declaration from a type node.
 */
export function findInterfaceFromTypeNode(
  typeNode: TypeNode,
  context: ExtractorContext,
): {
  declaration: InterfaceDeclaration;
  target:
    | { kind: "local"; filePath: string; name: string }
    | { kind: "module"; name: string };
} | null {
  if (!Node.isTypeReference(typeNode)) {
    return null;
  }

  const typeName = getTypeReferenceName(typeNode);
  const project = context.getProject();
  const sourceFile = context.getSourceFile();

  // Check current file first
  const localInterface = sourceFile
    .getInterfaces()
    .find((iface) => iface.getName() === typeName);
  if (localInterface) {
    return {
      declaration: localInterface,
      target: {
        kind: "local",
        filePath: sourceFile.getFilePath(),
        name: typeName,
      },
    };
  }

  // Search all project files
  for (const file of project.getSourceFiles()) {
    const interfaceDecl = file
      .getInterfaces()
      .find((iface) => iface.getName() === typeName);
    if (interfaceDecl) {
      return {
        declaration: interfaceDecl,
        target: { kind: "local", filePath: file.getFilePath(), name: typeName },
      };
    }
  }

  return null;
}

/**
 * Safely get the type name from a type reference node.
 */
export function getTypeReferenceName(typeNode: TypeReferenceNode): string {
  try {
    return typeNode.getTypeName().getText();
  } catch {
    return "";
  }
}

/**
 * Extract type arguments from a type reference node.
 */
export function parseTypeArguments(typeNode: TypeReferenceNode): TypeNode[] {
  try {
    return typeNode.getTypeArguments();
  } catch {
    return [];
  }
}

/**
 * Get generic type parameter names from a declaration.
 */
export function getGenericTypeParameters(
  declaration: InterfaceDeclaration | TypeAliasDeclaration,
): string[] {
  try {
    return declaration.getTypeParameters().map((param: any) => param.getName());
  } catch {
    return [];
  }
}

/**
 * Get the constraint or default type for a generic type parameter.
 * For example, in `T extends Asset = Asset`, returns the constraint `Asset`.
 */
export function getTypeParameterConstraintOrDefault(
  declaration: InterfaceDeclaration | TypeAliasDeclaration,
  parameterName: string,
): TypeNode | null {
  try {
    const typeParameters = declaration.getTypeParameters();
    const param = typeParameters.find(
      (p: any) => p.getName() === parameterName,
    );

    if (!param) return null;

    // Try default first, then constraint
    return param.getDefault() || param.getConstraint() || null;
  } catch {
    return null;
  }
}

/**
 * Handle nested utility type wrappers like Required<NonNullable<T>>.
 */
export function unwrapUtilityTypes(
  typeNode: TypeNode,
  wrapperTypes: string[] = [...UTILITY_TYPES],
): TypeNode {
  let currentType = typeNode;

  while (Node.isTypeReference(currentType)) {
    const typeName = getTypeReferenceName(currentType);

    if (wrapperTypes.includes(typeName)) {
      const typeArgs = currentType.getTypeArguments();
      if (typeArgs.length === 1) {
        currentType = typeArgs[0]!;
        continue;
      }
    }
    break;
  }

  return currentType;
}

/**
 * Check if a type node represents an array type (either T[] or Array<T>).
 */
export function isArrayType(typeNode: TypeNode): boolean {
  if (Node.isArrayTypeNode(typeNode)) {
    return true;
  }

  if (Node.isTypeReference(typeNode)) {
    const typeName = getTypeReferenceName(typeNode);
    return ARRAY_TYPE_NAMES.includes(typeName as any);
  }

  return false;
}

/**
 * Resolve generic parameters to their default types for interfaces without explicit type arguments.
 * Returns a map of generic parameter names to their default TypeNode values.
 */
export function resolveGenericParametersToDefaults(
  declaration: InterfaceDeclaration | TypeAliasDeclaration,
): Map<string, TypeNode> {
  const substitutions = new Map<string, TypeNode>();

  try {
    const typeParameters = declaration.getTypeParameters();

    for (const param of typeParameters) {
      const paramName = param.getName();

      // Try to get the default value first, then fall back to constraint
      const defaultType = param.getDefault() || param.getConstraint();

      if (defaultType) {
        substitutions.set(paramName, defaultType);
      }
    }
  } catch (error) {
    // If anything fails, return empty map to avoid breaking analysis
    console.warn(`Error resolving generic parameter defaults:`, error);
    return new Map();
  }

  return substitutions;
}

/**
 * Get the element type from an array type.
 */
export function getArrayElementType(typeNode: TypeNode): TypeNode | null {
  if (Node.isArrayTypeNode(typeNode)) {
    return typeNode.getElementTypeNode();
  }

  if (Node.isTypeReference(typeNode)) {
    const typeName = getTypeReferenceName(typeNode);
    if (ARRAY_TYPE_NAMES.includes(typeName as any)) {
      const typeArgs = typeNode.getTypeArguments();
      return typeArgs.length > 0 ? typeArgs[0]! : null;
    }
  }

  return null;
}
