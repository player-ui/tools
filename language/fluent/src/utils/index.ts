export * from "./set-at-path";
export * from "./tagged-template-handling";
export * from "./mark-as-builder";
export * from "./context";
