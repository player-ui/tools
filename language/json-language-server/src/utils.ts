import type { CancellationToken } from 'vscode-languageserver';
import { ResponseError } from 'vscode-languageserver';

/** Get a cancellation error   */
function cancel() {
  return new ResponseError(
    -32800 /* ErrorCodes.RequestCancelled */,
    'Request cancelled'
  );
}

/** Run the given function and handle being cancelled */
export async function runAndCatch<T>(
  func: () => Promise<T> | T,
  token: CancellationToken,
  errorVal: T
): Promise<T | ResponseError<any>> {
  if (token.isCancellationRequested) {
    return cancel();
  }

  try {
    const result = await func();
    if (token.isCancellationRequested) {
      return cancel();
    }

    return result;
  } catch (e) {
    return errorVal;
  }
}
