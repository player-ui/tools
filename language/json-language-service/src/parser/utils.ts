import type {
  ASTNode,
  PropertyASTNode,
  StringASTNode,
  ViewASTNode,
  ContentASTNode,
  FlowStateASTNode,
  ObjectASTNode,
  FlowASTNode,
  NavigationASTNode,
  AssetASTNode,
} from './types';

/** check if the node is a property */
export function isPropertyNode(node?: ASTNode): node is PropertyASTNode {
  return node?.type === 'property';
}

/** check if the node is an object ast */
export function isObjectNode(node?: ASTNode): node is ObjectASTNode {
  return node?.type === 'object';
}

/** check if the node is a key */
export function isKeyNode(node?: ASTNode): node is StringASTNode {
  return isPropertyNode(node?.parent) && node?.parent.keyNode === node;
}

/** check if the node is a value */
export function isValueNode(node?: ASTNode): boolean {
  return isPropertyNode(node?.parent) && node?.parent.valueNode === node;
}

/** check if the node is a view */
export function isViewNode(node?: ASTNode): node is ViewASTNode {
  return node?.type === 'view';
}

/** check if the node is a state */
export function isStateNode(node?: ASTNode): node is FlowStateASTNode {
  return node?.type === 'state';
}

/** check if the node is a flow */
export function isFlowNode(node?: ASTNode): node is FlowASTNode {
  return node?.type === 'flow';
}

/** check if the node is a nav */
export function isNavigationNode(node?: ASTNode): node is NavigationASTNode {
  return node?.type === 'navigation';
}

/** get the value for a given key in the object */
export function getValForKey<T extends ObjectASTNode<any>>(
  node: T,
  key: string
): any {
  const prop = node.properties.find((k) => k.keyNode.value === key);
  return prop?.valueNode?.jsonNode.value;
}

/** get the key for the given value node */
export function getPropForValNode(node?: ASTNode): string | undefined {
  if (isValueNode(node) && isPropertyNode(node?.parent)) {
    return node?.parent.keyNode.value;
  }
}

/** find the view node */
export function getViewNode(
  node: ASTNode
): ViewASTNode | AssetASTNode | undefined {
  if (node.parent) {
    if (node.parent.type === 'view') {
      return node.parent;
    }

    return getViewNode(node.parent);
  }

  if (node.type === 'asset') {
    return node;
  }
}

/** traverse up until we find the root content node */
export function getContentNode(node: ASTNode): ContentASTNode | undefined {
  if (node.type === 'content') {
    return node;
  }

  return node.parent ? getContentNode(node.parent) : undefined;
}
