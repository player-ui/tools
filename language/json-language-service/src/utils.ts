import { Range, Location } from 'vscode-languageserver-types';
import { TextDocument } from 'vscode-languageserver-textdocument';
import detectIndent from 'detect-indent';
import type {
  ASTNode,
  PlayerContent,
  ObjectASTNode,
  PropertyASTNode,
} from './parser';
import type { ASTVisitor } from './types';

export const typeToVisitorMap: Record<ASTNode['type'], keyof ASTVisitor> = {
  string: 'StringNode',
  number: 'NumberNode',
  boolean: 'BooleanNode',
  array: 'ArrayNode',
  null: 'NullNode',
  empty: 'EmptyNode',
  property: 'PropertyNode',
  object: 'ObjectNode',
  asset: 'AssetNode',
  view: 'ViewNode',
  content: 'ContentNode',
  navigation: 'NavigationNode',
  flow: 'FlowNode',
  state: 'FlowStateNode',
};

/** Check to see if the source range contains the target one */
export function containsRange(source: Range, range: Range): boolean {
  const { start, end } = range;
  const { start: srcStart, end: srcEnd } = source;

  return (
    start.line === srcStart.line &&
    end.line === srcEnd.line &&
    start.character >= srcStart.character &&
    end.character <= srcEnd.character
  );
}

/** Create a dummy TextDocument from the given string */
export function toTextDocument(str: string): TextDocument {
  return TextDocument.create('foo', 'json', 1, str);
}

/** Check to see if the document successfully parsed into a known root type */
export function isKnownRootType(document: PlayerContent): boolean {
  const { type } = document.root;

  return type === 'view' || type === 'asset' || type === 'content';
}

/** Check to see if the node is the value of an object */
export function isValueCompletion(node: ASTNode): boolean {
  return node.parent?.type === 'property' && node.parent.valueNode === node;
}

/** Check to see if the node is the key of an object */
export function isPropertyCompletion(node: ASTNode): boolean {
  return node.parent?.type === 'property' && node.parent.keyNode === node;
}

/** Search the object for a property with the given name */
export function getProperty<T extends ASTNode>(
  obj: T,
  name: string
): PropertyASTNode | undefined {
  if ('properties' in obj) {
    return (obj as ObjectASTNode).properties.find(
      (p) => p.keyNode.value === name
    );
  }
}

/** Get the LSP Location of an AST node in a document */
export function getLSLocationOfNode(
  document: TextDocument,
  node: ASTNode
): Location {
  const nodeRange = Range.create(
    document.positionAt(node.jsonNode.offset),
    document.positionAt(node.jsonNode.offset + node.jsonNode.length)
  );

  return Location.create(document.uri, nodeRange);
}

/** Get the depth of the property */
function getDepth(node: ASTNode): number {
  if (!node.parent) {
    return 0;
  }

  if (node.type === 'property') {
    return getDepth(node.parent);
  }

  return 1 + getDepth(node.parent);
}

/** Format a new node like an existing one (including the indentation) */
export function formatLikeNode(
  document: TextDocument,
  originalNode: ASTNode,
  replacement: Record<string, unknown>
): string {
  const { indent } = detectIndent(document.getText());
  const depth = getDepth(originalNode);

  return JSON.stringify(replacement, null, indent)
    .split('\n')
    .map((l, index) => (index === 0 ? l : `${indent.repeat(depth)}${l}`))
    .join('\n');
}

/** Maps the string identifying the FlowType to the named type */
export function mapFlowStateToType(flowType: string | undefined) {
  let flowXLR;
  switch (flowType) {
    case 'VIEW':
      flowXLR = 'NavigationFlowViewState';
      break;
    case 'END':
      flowXLR = 'NavigationFlowEndState';
      break;
    case 'ACTION':
      flowXLR = 'NavigationFlowActionState';
      break;
    case 'EXTERNAL':
      flowXLR = 'NavigationFlowExternalState';
      break;
    case 'FLOW':
      flowXLR = 'NavigationFlowFlowState';
      break;
    default:
      break;
  }

  return flowXLR;
}
