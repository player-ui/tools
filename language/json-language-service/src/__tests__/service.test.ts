import { TextDocument } from 'vscode-languageserver-textdocument';
import { PlayerLanguageService } from '..';
import { toTextDocument } from '../utils';

describe('player language service', () => {
  let service: PlayerLanguageService;

  beforeEach(async () => {
    service = new PlayerLanguageService();
    await service.setAssetTypes([
      './common/static_xlrs/core',
      './common/static_xlrs/plugin',
    ]);
  });

  describe('formatting', () => {
    it('formats a document', async () => {
      const document = toTextDocument(
        `{
"id": "foo",
"views": [{"id":"view1"}]
}
`
      );

      const formatEdits = await service.formatTextDocument(document, {
        tabSize: 2,
        insertSpaces: true,
      });
      expect(formatEdits).toHaveLength(8);
      const updatedDocument = TextDocument.applyEdits(
        document,
        formatEdits ?? []
      );
      expect(updatedDocument).toMatchInlineSnapshot(`
        "{
          \\"id\\": \\"foo\\",
          \\"views\\": [
            {
              \\"id\\": \\"view1\\"
            }
          ]
        }"
      `);
    });
  });

  describe('validation', () => {
    it('throws AssetWrapper errors', async () => {
      const document = toTextDocument(
        `
        {
          "id": "foo",
          "views": [
            {
              "id": "input-asset",
              "type": "input",
              "binding": "foo",
              "label": [{
                "id":"test",
                "type": "text",
                "value": "test"
              }]
            }
          ],
          "data": {},
          "navigation": {},
          "schema": {
            "ROOT": {
              "application": {
                "type": null
              },
              "foo": {
                "type": "fooType"
              }
            },
            "fooType": {
              "bar": {
                "type": "barType",
                "validation": [
                  {
                    "type": "required",
                    "severity": "error"
                  }
                ]
              },
              "baz": {
                "type": "bazType",
                "validation": [
                  {
                    "type": "required",
                    "message": "this is my own message"
                  }
                ]
              }
            }
          }
        }
        `
      );
      const validationErrors = await service.validateTextDocument(document);

      expect(validationErrors).toMatchSnapshot();
    });
  });

  describe('completion', () => {
    it('basic object completions', async () => {
      const document = toTextDocument(
        `
{
"id": "foo",
"views": [
  {
   "id": "view-1",
   "type": "main",
   "title": {
    "asset": {
     "id": "title-asset",
     "type": "text",
      
    }
   }
  }
],
"data":{},
"navigation":{

}
}
`
      );

      let completions = await service.getCompletionsAtPosition(document, {
        line: 11,
        character: 6,
      });
      expect(completions.items).toMatchSnapshot();

      completions = await service.getCompletionsAtPosition(document, {
        line: 18,
        character: 2,
      });
      expect(completions.items).toMatchSnapshot();
    });

    it('basic value completions', async () => {
      const document = toTextDocument(
        `
{
 "id": "foo",
 "views": [],
 "data":{},
 "navigation":{
  "BEGIN": "FLOW_1",
   "FLOW_1": {
   "startState": "VIEW_1",
   "VIEW_1": {
    "state_type": "",
    "ref": "action",
    "transitions": {
     "*": "END_Done"
    }
   },
   "END_Done": {
    "state_type": "END",
     "outcome": "done"
   }
  }
 }
}
`
      );

      const completions = await service.getCompletionsAtPosition(document, {
        line: 10,
        character: 19,
      });
      expect(completions.items).toMatchSnapshot();
    });

    it('schema completions', async () => {
      const document = toTextDocument(
        `
        {
          "id": "foo",
          "views": [
            {
              "id": "view-1",
              "type": "main",
              "title": {
                "asset": {
                  "id": "input-asset",
                  "type": "input",
                  "binding": "foo"
                }
              }
            }
          ],
          "data": {},
          "navigation": {},
          "schema": {
            "ROOT": {
              "application": {
                "type": null
              },
              "foo": {
                "type": "fooType"
              }
            },
            "fooType": {
              "bar": {
                "type": "barType",
                "validation": [
                  {
                    "type": "required",
                    "severity": "error"
                  }
                ]
              },
              "baz": {
                "type": "bazType",
                "validation": [
                  {
                    "type": "required",
                    "message": "this is my own message"
                  }
                ]
              }
            }
          }
        }
        
`
      );

      const completions = await service.getCompletionsAtPosition(document, {
        line: 11,
        character: 33,
      });
      expect(completions.items).toMatchSnapshot();
    });
  });

  describe('hover', () => {
    it('basic hover docs', async () => {
      const document = toTextDocument(
        `
{
  "id": "foo",
  "views": [
    {
      "id": "view-1",
      "type": "main",
      "title": {
        "asset": {
          "id": "title-asset",
          "type": "text"
        }
      }
    }
  ],
  "data": {},
  "navigation": {
    "BEGIN":""
  }
}
`
      );

      let docs = await service.getHoverInfoAtPosition(document, {
        line: 5,
        character: 9,
      });
      expect(docs).toBeDefined();
      expect(docs).toMatchSnapshot();

      docs = await service.getHoverInfoAtPosition(document, {
        line: 10,
        character: 13,
      });
      expect(docs).toBeDefined();
      expect(docs).toMatchSnapshot();

      docs = await service.getHoverInfoAtPosition(document, {
        line: 15,
        character: 6,
      });
      expect(docs).toBeDefined();
      expect(docs).toMatchSnapshot();
    });
  });
});
