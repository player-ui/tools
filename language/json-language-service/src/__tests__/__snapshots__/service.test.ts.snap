// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`player language service > completion > basic object completions 1`] = `
[
  {
    "documentation": "The text to display",
    "kind": 10,
    "label": "value",
  },
  {
    "documentation": "Any modifiers on the text",
    "kind": 10,
    "label": "modifiers",
  },
  {
    "documentation": "Evaluate the given expression (or boolean) and if falsy, remove this node from the tree. This is re-computed for each change in the data-model",
    "kind": 10,
    "label": "applicability",
  },
  {
    "documentation": "Adds a comment for the given node",
    "kind": 10,
    "label": "_comment",
  },
  {
    "documentation": "A list of templates to process for this node",
    "kind": 10,
    "label": "template",
  },
]
`;

exports[`player language service > completion > basic object completions 2`] = `
[
  {
    "documentation": "The name of the Flow to begin on",
    "kind": 10,
    "label": "BEGIN",
  },
]
`;

exports[`player language service > completion > basic value completions 1`] = `
[
  {
    "kind": 12,
    "label": "VIEW",
  },
  {
    "kind": 12,
    "label": "END",
  },
  {
    "kind": 12,
    "label": "FLOW",
  },
  {
    "kind": 12,
    "label": "ACTION",
  },
  {
    "kind": 12,
    "label": "EXTERNAL",
  },
]
`;

exports[`player language service > completion > schema completions 1`] = `
[
  {
    "kind": 12,
    "label": "",
  },
  {
    "kind": 12,
    "label": ".bar",
  },
  {
    "kind": 12,
    "label": ".baz",
  },
]
`;

exports[`player language service > hover > basic hover docs 1`] = `
{
  "contents": {
    "kind": "plaintext",
    "value": "Each asset requires a unique id per view",
  },
}
`;

exports[`player language service > hover > basic hover docs 2`] = `
{
  "contents": {
    "kind": "plaintext",
    "value": "The asset type determines the semantics of how a user interacts with a page",
  },
}
`;

exports[`player language service > hover > basic hover docs 3`] = `
{
  "contents": {
    "kind": "plaintext",
    "value": "Any initial data that the flow can use",
  },
}
`;

exports[`player language service > validation > throws AssetWrapper errors 1`] = `
[
  {
    "message": "Content Validation Error - type: Expected type "string" but got "null"",
    "range": {
      "end": {
        "character": 28,
        "line": 20,
      },
      "start": {
        "character": 16,
        "line": 20,
      },
    },
    "severity": 1,
  },
  {
    "message": "Content Validation Error - missing: Property "BEGIN" missing from type "Navigation"",
    "range": {
      "end": {
        "character": 26,
        "line": 16,
      },
      "start": {
        "character": 24,
        "line": 16,
      },
    },
    "severity": 1,
  },
  {
    "message": "Navigation Validation Error - missing: Property "BEGIN" missing from type "Navigation"",
    "range": {
      "end": {
        "character": 26,
        "line": 16,
      },
      "start": {
        "character": 24,
        "line": 16,
      },
    },
    "severity": 1,
  },
  {
    "message": "View is not reachable",
    "range": {
      "end": {
        "character": 33,
        "line": 5,
      },
      "start": {
        "character": 20,
        "line": 5,
      },
    },
    "severity": 2,
  },
  {
    "message": "View Validation Error - value: Does not match any of the expected types for type: 'AssetWrapperOrSwitch'",
    "range": {
      "end": {
        "character": 16,
        "line": 12,
      },
      "start": {
        "character": 23,
        "line": 8,
      },
    },
    "severity": 1,
  },
  {
    "message": "Implicit Array -> "collection" assets is not supported.",
    "range": {
      "end": {
        "character": 21,
        "line": 8,
      },
      "start": {
        "character": 14,
        "line": 8,
      },
    },
    "severity": 1,
  },
]
`;
